### PR TITLE
refactor: thrift codec uses bufiox interface for encoding and decoding

### PR DESCRIPTION
**Pull request by jayantxie**
#### What type of PR is this?
refactor

#### Check the PR title.

- [ ] This PR title match the format: \<type\>(optional scope): \<description\>
- [ ] The description of this PR title is user-oriented and clear enough for others to understand.
- [ ] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en:
zh(optional):

### CODE REVIEW

File: protocol/thrift/binarywriter.go [hunk 1]
```diff
@@ -1,120 +0,0 @@
...
-package thrift
-
-import (
-	"math"
-	"sync"
-)
-
-const defaultBinaryWriterBufferSize = 4096
-
-type BinaryWriter struct {
```
Comment (xiaost) [inline protocol/thrift/binarywriter.go:26]: 这个泛化那边在用好像，之前专门给那边用来拼bytes的

File: protocol/thrift/bufferwriter_test.go [hunk 2]
```diff
@@ -19,68 +19,63 @@ package thrift
...
+
 func TestBinaryWriter(t *testing.T) {
-	w := NewBinaryWriterSize(defaultBinaryWriterBufferSize * 2)
+	buf := dirtmake.Bytes(0, defaultBinaryWriterBufferSize*2)
+	w := NewBufferWriter(bufiox.NewBytesWriter(&buf))
 	x := BinaryProtocol{}
 
 	b := x.AppendMessageBegin(nil, "hello", 1, 2)
 	w.WriteMessageBegin("hello", 1, 2)
-	require.Equal(t, b, w.Bytes())
```
Comment (xiaost) [inline protocol/thrift/bufferwriter_test.go:31]: 这个是为什么删了么？之前专门加来DEBUG是哪个编码开始有问题的。

File: protocol/thrift/binarywriter.go (hunk 1 above)
Comment (jayantxie) [reply protocol/thrift/binarywriter.go:26]: 我改成了BufferWriter，传入的是bufiox.Writer，去掉了buffer拼接的接口。

File: protocol/thrift/bufferwriter_test.go (hunk 2 above)
Comment (jayantxie) [reply protocol/thrift/bufferwriter_test.go:31]: 这个我改成在最后校验了，因为没有flush之前无法直接对比，BufferWriter没有提供Bytes方法

**Review by YangruiEmma — APPROVED** [review]

**Pull request merged**